crates/services/txpool_v2/src/tests/pool.rs

```diff
@@ -0,0 +1,1582 @@
+    })
+    .add_random_fee_input()
+    .add_witness(program.into())
+    .finalize_as_transaction();
+    let same_blob_tx = check_unwrap_tx(same_blob_tx, &txpool.config).await;
+
+    // When
+    let results = txpool
+        .insert(vec![check_tx_to_pool(same_blob_tx.clone())])
+        .unwrap();
+    assert_eq!(results.len(), 1);
+    let err = results[0].as_ref().expect_err("Tx should be Err, got Ok");
+
+    // Then
+    assert!(matches!(err, Error::Collided(_)));
+}
+
+#[tokio::test]
+async fn insert_single__blob_tx_succeeds_if_blob_already_inserted_but_higher_tip() {
+    let program = vec![123; 123];
+    let blob_id = BlobId::compute(program.as_slice());
+    let tx = TransactionBuilder::blob(BlobBody {
+        id: blob_id,
+        witness_index: 0,
+    })
+    .add_witness(program.clone().into())
+    .add_random_fee_input()
+    .finalize_as_transaction();
+
+    let config = Config {
+        utxo_validation: false,
+        ..Default::default()
+    };
+    let context = PoolContext::default().config(config);
+    let mut txpool = context.build();
+    let tx = check_unwrap_tx(tx, &txpool.config).await;
+
+    // Given
+    let results = txpool.insert(vec![check_tx_to_pool(tx.clone())]).unwrap();
+    assert_eq!(results.len(), 1);
+    assert!(results[0].is_ok());
+
+    let same_blob_tx = TransactionBuilder::blob(BlobBody {
+        id: blob_id,
+        witness_index: 1,
+    })
+    .add_random_fee_input()
+    .add_witness(program.into())
+    .tip(100)
+    .max_fee_limit(100)
+    .finalize_as_transaction();
+    let same_blob_tx = check_unwrap_tx(same_blob_tx, &txpool.config).await;
+
+    // When
+    let results = txpool
+        .insert(vec![check_tx_to_pool(same_blob_tx.clone())])
+        .unwrap();
+    assert_eq!(results.len(), 1);
+
+    // Then
+    assert!(results[0].is_ok());
+}
+
+#[tokio::test]
+async fn insert_single__blob_tx_fails_if_blob_already_exists_in_database() {
+    let program = vec![123; 123];
+    let blob_id = BlobId::compute(program.as_slice());
+    let tx = TransactionBuilder::blob(BlobBody {
+        id: blob_id,
+        witness_index: 0,
+    })
+    .add_witness(program.clone().into())
+    .add_random_fee_input()
+    .finalize_as_transaction();
+
+    let config = Config {
+        utxo_validation: false,
+        ..Default::default()
+    };
+    let mut context = PoolContext::default().config(config);
+    // Given
+    context.database_mut().insert_dummy_blob(blob_id);
+    let mut txpool = context.build();
+    let tx = check_unwrap_tx(tx, &txpool.config).await;
+
+    // When
+    let result = txpool.insert(vec![check_tx_to_pool(tx.clone())]).unwrap();
+    assert_eq!(result.len(), 1);
+    let err = result[0].as_ref().expect_err("Tx should be Err, got Ok");
+
+    // Then
+    assert!(matches!(
+        err,
+        Error::NotInsertedBlobIdAlreadyTaken(b) if *b == blob_id
+    ));
+}
+
+#[tokio::test]
+async fn remove_dependency_chain_on_insertion() {
```

Comment:
Can we specify the SUT and give a G/W/T as well? Also "remove dependency chain" of what "on insertion" of what?

Reply:
Done.